pkg/services/provisioning/datasources/config_reader.go

```diff
@@ -51,35 +51,50 @@ func (cr *configReader) parseDatasourceConfig(path string, file os.FileInfo) (*D
 		return nil, err
 	}
 
-	var apiVersion *ConfigVersion
-	err = yaml.Unmarshal(yamlFile, &apiVersion)
+	apiVersion, err := getDatasourceConfigApiVersion(yamlFile)
 	if err != nil {
 		return nil, err
 	}
 
-	if apiVersion == nil {
-		apiVersion = &ConfigVersion{ApiVersion: 0}
+	if apiVersion == 0 {
+		cr.log.Warn("[Deprecated] the datasource provisioning config is outdated. please upgrade", "filename", filename)
 	}
 
-	if apiVersion.ApiVersion > 0 {
+	return cr.getDatasourceAsConfigForVersion(apiVersion, yamlFile)
+}
+
+func (cr *configReader) getDatasourceAsConfigForVersion(version int64, yamlFile []byte) (*DatasourcesAsConfig, error) {
+	if version > 0 {
 		var v1 *DatasourcesAsConfigV1
-		err = yaml.Unmarshal(yamlFile, &v1)
+		err := yaml.Unmarshal(yamlFile, &v1)
 		if err != nil {
 			return nil, err
 		}
 
-		return v1.mapToDatasourceFromConfig(apiVersion.ApiVersion), nil
+		return v1.mapToDatasourceFromConfig(version), nil
+	} else {
+		var v0 *DatasourcesAsConfigV0
+		err := yaml.Unmarshal(yamlFile, &v0)
+		if err != nil {
+			return nil, err
+		}
+
+		return v0.mapToDatasourceFromConfig(version), nil
 	}
+}
 
-	var v0 *DatasourcesAsConfigV0
-	err = yaml.Unmarshal(yamlFile, &v0)
+func getDatasourceConfigApiVersion(yamlFile []byte) (int64, error) {
+	var apiVersion *ConfigVersion
+	err := yaml.Unmarshal(yamlFile, &apiVersion)
 	if err != nil {
-		return nil, err
+		return -1, err
 	}
 
-	cr.log.Warn("[Deprecated] the datasource provisioning config is outdated. please upgrade", "filename", filename)
+	if apiVersion == nil {
+		return 0, nil
```

Comment:
Shouldn't this be 1 per default? On another note maybe better to do config_reader refactoring in a separate PR to make this PR smaller?

Reply:
As I mentioned will do a separate PR with this but I think this is correct, in the previous code it was:
```go
    if apiVersion == nil {
		apiVersion = &ConfigVersion{ApiVersion: 0}
	}

	if apiVersion.ApiVersion > 0 {
		var v1 *DatasourcesAsConfigV1
```
so nil => 0, 0 => 0, > 0 => 1

Reply:
Oh okay :+1: